src/recordlayer.js

```diff
@@ -0,0 +1,324 @@
+//    // until you specify the key to use.
+//    await rl.setSendKey(key)
+//
+//    // Send some data by specifying the record type and the bytes.
+//    // Where allowed by the record type, it will be buffered until
+//    // explicitly flushed, and then sent by calling the callback.
+//    await rl.send(RECORD_TYPE.HANDSHAKE, <bytes for a handshake message>)
+//    await rl.send(RECORD_TYPE.HANDSHAKE, <bytes for another handshake message>)
+//    await rl.flush()
+//
+//    // Separate keys are used for sending and receiving.
+//    rl.setRecvKey(key);
+//
+//    // When data is received, push it into the RecordLayer
+//    // and pass a callback that will be called with a [type, bytes]
+//    // pair for each message parsed from the data.
+//    rl.recv(dataReceivedFromPeer, async (type, bytes) => {
+//      switch (type) {
+//        case RECORD_TYPE.APPLICATION_DATA:
+//          // do something with application data
+//        case RECORD_TYPE.HANDSHAKE:
+//          // do something with a handshake message
+//        default:
+//          // etc...
+//      }
+//    });
+//
+
+
+import { VERSION_TLS_1_2 } from './constants.js';
+import {
+  encrypt,
+  decrypt,
+  prepareKey,
+  hkdfExpandLabel,
+  AEAD_SIZE_INFLATION,
+  IV_LENGTH,
+  KEY_LENGTH,
+} from './crypto.js';
+import {
+  assert,
+  assertIsBytes,
+  BufferReader,
+  BufferWriter,
+  EMPTY,
+} from './utils.js';
+import { alertIf, ALERT_DESCRIPTION } from './alerts.js';
+
+/* eslint-disable sorting/sort-object-props */
+export const RECORD_TYPE = {
+  CHANGE_CIPHER_SPEC: 20,
+  ALERT: 21,
+  HANDSHAKE: 22,
+  APPLICATION_DATA: 23,
+};
+/* eslint-enable sorting/sort-object-props */
+
+// Encrypting at most 2^24 records will force us to stay
+// below data limits on AES-GCM encryption key use, and also
+// means we can accurately represent the sequence number as
+// a javascript double.
+const MAX_SEQUENCE_NUMBER = Math.pow(2, 24);
+const MAX_RECORD_SIZE = Math.pow(2, 14);
+const MAX_ENCRYPTED_RECORD_SIZE = MAX_RECORD_SIZE + 256;
+const RECORD_HEADER_SIZE = 5;
+
+// This is a helper class to manage the encryption/decryption state
+// for a particular key.
+
+class CipherState {
+  constructor() {
+    this.key = null;
+    this.iv = null;
+    this.seqnum = 0;
+  }
+
+  async setKey(key, mode) {
+    assertIsBytes(key);
+    // Derive key and iv per https://tools.ietf.org/html/rfc8446#section-7.3
+    this.key = await prepareKey(await hkdfExpandLabel(key, 'key', EMPTY, KEY_LENGTH), mode);
+    this.iv = await hkdfExpandLabel(key, 'iv', EMPTY, IV_LENGTH);
+    this.seqnum = 0;
+  }
+
+  nonce() {
+    // Ref https://tools.ietf.org/html/rfc8446#section-5.3:
+    // * left-pad the sequence number with zeros to IV_LENGTH
+    // * xor with the provided iv
+    const nonce = new Uint8Array(IV_LENGTH);
+    // Our sequence numbers are always less than 2^24, so fit in a Uint32.
+    (new DataView(nonce.buffer)).setUint32(IV_LENGTH - 4, this.seqnum);
```

Comment:
`new DataView(nonce.buffer).setUint32(IV_LENGTH - 4, this.seqnum);` ? remove extra `(`?